File: app/src/main/java/com/example/papercamera/SettingsDialogController.java
package com.domonation.camera;

import android.app.Dialog;
import android.content.SharedPreferences;
import android.graphics.Color;
import android.text.Editable;
import android.text.InputType;
import android.text.TextWatcher;
import android.view.Gravity;
import android.view.KeyEvent;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.CheckBox;
import android.widget.EditText;
import android.widget.ImageButton;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.RadioButton;
import android.widget.RadioGroup;
import android.widget.ScrollView;
import android.widget.TextView;

import androidx.activity.ComponentActivity;
import androidx.core.content.ContextCompat;

final class SettingsDialogController {
    private static final String KEY_TREE = "save_tree";
    private static final String KEY_REVIEW = "review_before_save";
    private static final String KEY_THUMB = "exif_thumbnail";
    private static final String KEY_LAPSE = "timelapse_seconds";
    private static final String KEY_LAPSE_FPS = "timelapse_fps";
    private static final String KEY_LAPSE_CUSTOM = "timelapse_seconds_custom";
    private static final String KEY_LAPSE_FPS_CUSTOM = "timelapse_fps_custom";
    private static final String KEY_VIDEO_AUDIO = "video_audio";
    private static final String KEY_VOLUME_SHUTTER = "volume_shutter";
    private static final String KEY_GALLERY_ROWS = "gallery_rows";
    private static final String KEY_GALLERY_COLUMNS = "gallery_columns";

    interface Host {
        void chooseFolder();
        void useDefaultFolder();
        void onSettingsDismissed();
    }

    private final ComponentActivity activity;
    private final SharedPreferences prefs;
    private final Host host;

    SettingsDialogController(ComponentActivity activity, SharedPreferences prefs, Host host) {
        this.activity = activity;
        this.prefs = prefs;
        this.host = host;
    }

    void show() {
        Dialog dialog = new Dialog(activity, android.R.style.Theme_Material_Light_NoActionBar_Fullscreen);
        LinearLayout page = new LinearLayout(activity);
        page.setOrientation(LinearLayout.VERTICAL);
        page.setBackgroundColor(color(R.color.paper));

        LinearLayout header = new LinearLayout(activity);
        header.setGravity(Gravity.CENTER_VERTICAL);
        header.setBackgroundResource(R.drawable.top_rule);
        Button close = textButton("×");
        close.setTextSize(30);
        header.addView(close, new LinearLayout.LayoutParams(dp(56), dp(56)));
        TextView title = label("Settings");
        title.setTextSize(21);
        title.setTypeface(null, android.graphics.Typeface.BOLD);
        title.setGravity(Gravity.CENTER_VERTICAL);
        title.setPadding(0, 0, 0, 0);
        header.addView(title, new LinearLayout.LayoutParams(0, dp(56), 1));
        page.addView(header, new LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, dp(56)));

        LinearLayout tabs = new LinearLayout(activity);
        tabs.setBackgroundResource(R.drawable.top_rule);
        String[] names = {"General settings", "Photo settings", "Video settings", "Timelapse settings"};
        int[] icons = {R.drawable.ic_settings, R.drawable.ic_photo,
                R.drawable.ic_video, R.drawable.ic_timelapse};
        ImageButton[] tabButtons = new ImageButton[names.length];
        LinearLayout content = new LinearLayout(activity);
        content.setOrientation(LinearLayout.VERTICAL);
        int[] selected = {0};
        for (int i = 0; i < names.length; i++) {
            ImageButton tab = tabButton(icons[i], names[i]);
            tab.setTag(i);
            tabButtons[i] = tab;
            tabs.addView(tab, new LinearLayout.LayoutParams(0, dp(48), 1));
            tab.setOnClickListener(v -> {
                selected[0] = (int) v.getTag();
                styleTabs(tabButtons, selected[0]);
                populate(content, selected[0], dialog);
            });
        }
        page.addView(tabs, new LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, dp(48)));

        ScrollView scroller = new ScrollView(activity);
        scroller.setFillViewport(true);
        scroller.addView(content, new ScrollView.LayoutParams(
                ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT));
        page.addView(scroller, new LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, 0, 1));
        close.setOnClickListener(v -> dialog.dismiss());
        dialog.setOnKeyListener((d, keyCode, event) -> {
            if (keyCode != KeyEvent.KEYCODE_BACK) return false;
            if (event.getAction() == KeyEvent.ACTION_UP) dialog.dismiss();
            return true;
        });
        dialog.setOnDismissListener(d -> host.onSettingsDismissed());
        dialog.setContentView(page);
        dialog.show();
        if (dialog.getWindow() != null) dialog.getWindow().setLayout(
                ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT);
        styleTabs(tabButtons, 0);
        populate(content, 0, dialog);
    }

    private void populate(LinearLayout content, int page, Dialog dialog) {
        content.removeAllViews();
        content.setPadding(dp(20), dp(2), dp(20), dp(18));
        if (page == 0) {
            content.addView(sectionTitle("General"));
            content.addView(checkSetting("Review before save", KEY_REVIEW, false));
            content.addView(checkSetting("Volume buttons control shutter", KEY_VOLUME_SHUTTER, true));
            content.addView(sectionTitle("Gallery grid"));
            content.addView(note("Rows per page"));
            content.addView(radioSetting(KEY_GALLERY_ROWS, prefs.getInt(KEY_GALLERY_ROWS, 3)));
            content.addView(note("Columns per page"));
            content.addView(radioSetting(KEY_GALLERY_COLUMNS, prefs.getInt(KEY_GALLERY_COLUMNS, 3)));
            content.addView(sectionTitle("Save location"));
            TextView location = note(prefs.getString(KEY_TREE, null) == null ?
                    "Pictures/PaperCamera and Movies/PaperCamera" : "Selected folder (photos and videos)");
            content.addView(location);
            Button choose = settingsButton("CHOOSE FOLDER");
            choose.setOnClickListener(v -> { dialog.dismiss(); host.chooseFolder(); });
            content.addView(choose, matchHeight(48));
            Button defaults = settingsButton("USE DEFAULT FOLDERS");
            defaults.setOnClickListener(v -> {
                prefs.edit().remove(KEY_TREE).apply();
                location.setText("Pictures/PaperCamera and Movies/PaperCamera");
                host.useDefaultFolder();
            });
            content.addView(defaults, matchHeight(48));
        } else if (page == 1) {
            content.addView(sectionTitle("Photo"));
            content.addView(checkSetting("Embed EXIF thumbnail", KEY_THUMB, true));
            content.addView(note("The embedded thumbnail improves browsing on devices such as the Canon SELPHY."));
        } else if (page == 2) {
            content.addView(sectionTitle("Video"));
            content.addView(checkSetting("Record audio", KEY_VIDEO_AUDIO, true));
            content.addView(note("Turn this off for silent video without microphone permission."));
        } else {
            content.addView(sectionTitle("Timelapse"));
            content.addView(note("Each captured JPEG is saved, followed by an MP4 made from the same frames."));
            content.addView(sectionTitle("Capture interval (sec)"));
            content.addView(customNumberSetting(KEY_LAPSE, KEY_LAPSE_CUSTOM,
                    new String[]{"2", "5", "10"}, new int[]{2, 5, 10}, prefs.getInt(KEY_LAPSE, 5),
                    15, 1, 86400, "Seconds (1–86400)"));
            content.addView(sectionTitle("Video playback rate (FPS)"));
            content.addView(customNumberSetting(KEY_LAPSE_FPS, KEY_LAPSE_FPS_CUSTOM,
                    new String[]{"12", "24", "30"}, new int[]{12, 24, 30}, prefs.getInt(KEY_LAPSE_FPS, 24),
                    25, 1, 60, "FPS (1–60)"));
        }
    }

    private CheckBox checkSetting(String text, String key, boolean defaultValue) {
        CheckBox box = new CheckBox(activity);
        box.setText(text);
        box.setTextColor(color(R.color.ink));
        box.setTextSize(16);
        box.setMinHeight(dp(44));
        box.setChecked(prefs.getBoolean(key, defaultValue));
        box.setOnCheckedChangeListener((button, checked) -> prefs.edit().putBoolean(key, checked).apply());
        return box;
    }

    private RadioGroup radioSetting(String key, int selectedValue) {
        return radioSetting(key, new String[]{"2", "3", "4"}, new int[]{2, 3, 4}, selectedValue);
    }

    private RadioGroup radioSetting(String key, String[] labels, int[] values, int selectedValue) {
        RadioGroup group = new RadioGroup(activity);
        group.setOrientation(RadioGroup.HORIZONTAL);
        int checkedId = View.NO_ID;
        for (int i = 0; i < labels.length; i++) {
            RadioButton option = radio(labels[i], values[i]);
            group.addView(option, new RadioGroup.LayoutParams(0, dp(44), 1f));
            if (values[i] == selectedValue) checkedId = option.getId();
        }
        if (checkedId != View.NO_ID) group.check(checkedId);
        group.setOnCheckedChangeListener((radioGroup, id) -> {
            RadioButton checked = radioGroup.findViewById(id);
            if (checked != null) prefs.edit().putInt(key, (int) checked.getTag()).apply();
        });
        return group;
    }

    private View customNumberSetting(String key, String customKey, String[] labels, int[] values,
                                     int selectedValue, int customDefault, int min, int max, String hint) {
        LinearLayout wrapper = new LinearLayout(activity);
        wrapper.setOrientation(LinearLayout.VERTICAL);
        RadioGroup group = new RadioGroup(activity);
        group.setOrientation(RadioGroup.HORIZONTAL);
        int checkedId = View.NO_ID;
        boolean preset = false;
        for (int i = 0; i < labels.length; i++) {
            RadioButton option = radio(labels[i], values[i]);
            group.addView(option, new RadioGroup.LayoutParams(0, dp(44), 1f));
            if (values[i] == selectedValue) { checkedId = option.getId(); preset = true; }
        }
        RadioButton custom = radio("Custom", null);
        group.addView(custom, new RadioGroup.LayoutParams(0, dp(44), 1.35f));
        if (!preset) checkedId = custom.getId();
        EditText entry = numberEntry(hint, prefs.getInt(customKey,
                preset ? customDefault : Math.max(min, Math.min(max, selectedValue))));
        entry.setVisibility(preset ? View.GONE : View.VISIBLE);
        group.check(checkedId);
        group.setOnCheckedChangeListener((radioGroup, id) -> {
            if (id == custom.getId()) {
                entry.setVisibility(View.VISIBLE);
                int value = validNumber(entry.getText().toString(), min, max,
                        prefs.getInt(customKey, customDefault));
                prefs.edit().putInt(key, value).putInt(customKey, value).apply();
            } else {
                RadioButton checked = radioGroup.findViewById(id);
                if (checked == null || !(checked.getTag() instanceof Integer)) return;
                entry.setVisibility(View.GONE);
                prefs.edit().putInt(key, (int) checked.getTag()).apply();
            }
        });
        entry.addTextChangedListener(new TextWatcher() {
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {}
            public void onTextChanged(CharSequence s, int start, int before, int count) {}
            public void afterTextChanged(Editable text) {
                if (group.getCheckedRadioButtonId() != custom.getId()) return;
                try {
                    int value = Integer.parseInt(text.toString());
                    if (value >= min && value <= max)
                        prefs.edit().putInt(key, value).putInt(customKey, value).apply();
                } catch (NumberFormatException ignored) { }
            }
        });
        wrapper.addView(group, matchHeight(44));
        wrapper.addView(entry, matchHeight(48));
        return wrapper;
    }

    private RadioButton radio(String text, Object tag) {
        RadioButton button = new RadioButton(activity);
        button.setId(View.generateViewId());
        button.setTag(tag);
        button.setText(text);
        button.setTextColor(color(R.color.ink));
        button.setTextSize(15);
        button.setMinHeight(dp(44));
        return button;
    }

    private EditText numberEntry(String hint, int value) {
        EditText entry = new EditText(activity);
        entry.setSingleLine(true);
        entry.setInputType(InputType.TYPE_CLASS_NUMBER);
        entry.setTextColor(color(R.color.ink));
        entry.setHintTextColor(color(R.color.mid_ink));
        entry.setTextSize(15);
        entry.setHint(hint);
        entry.setContentDescription("Custom " + hint);
        entry.setSelectAllOnFocus(true);
        entry.setText(String.valueOf(value));
        return entry;
    }

    private int validNumber(String text, int min, int max, int fallback) {
        try { return Math.max(min, Math.min(max, Integer.parseInt(text))); }
        catch (NumberFormatException ignored) { return Math.max(min, Math.min(max, fallback)); }
    }

    private void styleTabs(ImageButton[] tabs, int selected) {
        for (int i = 0; i < tabs.length; i++) {
            boolean active = i == selected;
            tabs[i].setBackgroundColor(color(active ? R.color.ink : R.color.paper));
            tabs[i].setImageTintList(android.content.res.ColorStateList.valueOf(
                    color(active ? R.color.paper : R.color.ink)));
        }
    }

    private ImageButton tabButton(int icon, String description) {
        ImageButton button = new ImageButton(activity);
        button.setImageResource(icon);
        button.setContentDescription(description);
        button.setScaleType(ImageView.ScaleType.CENTER_INSIDE);
        button.setPadding(dp(14), dp(10), dp(14), dp(10));
        button.setBackgroundColor(Color.TRANSPARENT);
        button.setStateListAnimator(null);
        return button;
    }

    private Button settingsButton(String text) {
        Button button = textButton(text);
        button.setTextSize(13);
        button.setBackgroundResource(R.drawable.top_rule);
        return button;
    }

    private Button textButton(String text) {
        Button button = new Button(activity);
        button.setText(text);
        button.setTextColor(color(R.color.ink));
        button.setTextSize(16);
        button.setBackgroundColor(Color.TRANSPARENT);
        button.setStateListAnimator(null);
        button.setMinimumHeight(0);
        button.setMinimumWidth(0);
        return button;
    }

    private TextView label(String text) {
        TextView view = new TextView(activity);
        view.setText(text);
        view.setTextColor(Color.BLACK);
        view.setTextSize(17);
        view.setPadding(0, 14, 0, 4);
        return view;
    }

    private TextView sectionTitle(String text) {
        TextView view = label(text);
        view.setTextSize(18);
        view.setTypeface(null, android.graphics.Typeface.BOLD);
        view.setPadding(0, dp(8), 0, dp(4));
        return view;
    }

    private TextView note(String text) {
        TextView view = label(text);
        view.setTextSize(14);
        view.setTextColor(color(R.color.mid_ink));
        view.setPadding(0, dp(2), 0, dp(8));
        return view;
    }

    private LinearLayout.LayoutParams matchHeight(int height) {
        return new LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, dp(height));
    }

    private int color(int id) { return ContextCompat.getColor(activity, id); }
    private int dp(int value) {
        return Math.round(value * activity.getResources().getDisplayMetrics().density);
    }
}

File: app/src/main/java/com/domonation/camera/SettingsDialogController.kt
package com.domonation.camera

import android.app.Dialog
import android.content.SharedPreferences
import android.graphics.drawable.ColorDrawable
import android.net.Uri
import android.provider.DocumentsContract
import android.view.KeyEvent
import android.view.ViewGroup
import androidx.activity.ComponentActivity
import androidx.activity.ComponentDialog
import androidx.core.view.WindowCompat
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.background
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.MaterialTheme
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.platform.ComposeView
import androidx.compose.ui.platform.ViewCompositionStrategy
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.input.KeyboardType
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.Scaffold
import com.mudita.mmd.ThemeMMD
import com.mudita.mmd.components.buttons.ButtonMMD
import com.mudita.mmd.components.buttons.OutlinedButtonMMD
import com.mudita.mmd.components.divider.HorizontalDividerMMD
import com.mudita.mmd.components.radio_button.RadioButtonMMD
import com.mudita.mmd.components.switcher.SwitchMMD
import com.mudita.mmd.components.nav_bar.NavigationBarItemMMD
import com.mudita.mmd.components.nav_bar.NavigationBarMMD
import com.mudita.mmd.components.text.TextMMD
import com.mudita.mmd.components.text_field.TextFieldMMD
import com.mudita.mmd.components.top_app_bar.TopAppBarMMD

internal class SettingsDialogController(
    private val activity: ComponentActivity,
    private val prefs: SharedPreferences,
    private val host: Host,
) {
    interface Host {
        fun chooseFolder()
        fun onSettingsDismissed()
    }

    fun show() {
        val dialog = ComponentDialog(activity, android.R.style.Theme_Material_Light_NoActionBar)
        val compose = ComposeView(activity).apply {
            setViewCompositionStrategy(ViewCompositionStrategy.DisposeOnDetachedFromWindow)
            setContent { ThemeMMD(colorScheme = MmdUi.domonationColorScheme) { SettingsScreen(dialog) } }
        }
        dialog.setOnKeyListener { _, keyCode, event ->
            if (keyCode == KeyEvent.KEYCODE_BACK && event.action == KeyEvent.ACTION_UP) {
                dialog.dismiss(); true
            } else false
        }
        dialog.setOnDismissListener { host.onSettingsDismissed() }
        dialog.setContentView(compose)
        dialog.show()
        dialog.window?.apply {
            WindowCompat.setDecorFitsSystemWindows(this, true)
            setBackgroundDrawable(ColorDrawable(android.graphics.Color.WHITE))
            statusBarColor = android.graphics.Color.WHITE
            navigationBarColor = android.graphics.Color.WHITE
            decorView.systemUiVisibility = android.view.View.SYSTEM_UI_FLAG_LIGHT_STATUS_BAR or
                android.view.View.SYSTEM_UI_FLAG_LIGHT_NAVIGATION_BAR
            setLayout(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT)
        }
    }

    @Composable
    @OptIn(ExperimentalMaterial3Api::class)
    private fun SettingsScreen(dialog: Dialog) {
        var tab by remember { mutableIntStateOf(0) }
        val destinations = listOf(
            Triple("General", R.drawable.ic_settings, "General settings"),
            Triple("Photo", R.drawable.ic_photo, "Photo settings"),
            Triple("Video", R.drawable.ic_video, "Video settings"),
            Triple("Timelapse", R.drawable.ic_timelapse, "Timelapse settings"),
        )
        Scaffold(
            modifier = Modifier.fillMaxSize(),
            containerColor = Color.White,
            topBar = {
                Box(Modifier.fillMaxWidth().height(64.dp)) {
                    TopAppBarMMD(
                        title = { TextMMD("Settings", fontSize = 28.sp, fontWeight = FontWeight.Bold) },
                        navigationIcon = {
                            IconButton(onClick = { dialog.dismiss() }) {
                                Icon(
                                    painter = painterResource(R.drawable.ic_back),
                                    contentDescription = "Close settings",
                                    tint = Color.Black,
                                    modifier = Modifier.size(44.dp),
                                )
                            }
                        },
                        showDivider = false,
                    )
                    HorizontalDividerMMD(color = Color.Black, thickness = 3.dp,
                        modifier = Modifier.align(Alignment.BottomCenter))
                }
            },
            bottomBar = {
                Column(Modifier.fillMaxWidth().height(88.dp).background(Color.White)) {
                    HorizontalDividerMMD(color = Color.Black, thickness = 3.dp)
                    NavigationBarMMD(
                        modifier = Modifier.weight(1f),
                        containerColor = Color.White,
                        contentColor = Color.Black,
                        windowInsets = WindowInsets(0),
                    ) {
                        destinations.forEachIndexed { index, (label, icon, description) ->
                            NavigationBarItemMMD(
                                selected = tab == index,
                                onClick = { tab = index },
                                icon = {
                                    Icon(
                                        painter = painterResource(icon),
                                        contentDescription = description,
                                        modifier = Modifier.size(24.dp),
                                    )
                                },
                                label = { TextMMD(label, fontWeight = if (tab == index) FontWeight.Bold else FontWeight.Normal) },
                            )
                        }
                    }
                }
            },
        ) { innerPadding ->
            Column(
                Modifier.fillMaxSize().padding(innerPadding).background(Color.White)
                    .verticalScroll(rememberScrollState()).padding(16.dp),
                verticalArrangement = Arrangement.spacedBy(2.dp),
            ) {
                when (tab) {
                    0 -> GeneralSettings(dialog)
                    1 -> PhotoSettings()
                    2 -> VideoSettings()
                    else -> TimelapseSettings()
                }
            }
        }
    }

    @Composable
    private fun GeneralSettings(dialog: Dialog) {
        Section("General")
        ToggleRow("Review before save", KEY_REVIEW, false)
        ToggleRow("Volume buttons control shutter", KEY_VOLUME_SHUTTER, true)
        ToggleRow("Flip front camera horizontally", KEY_FLIP_FRONT_CAMERA, true)
        Section("Save location")
        ButtonMMD(
            onClick = { dialog.dismiss(); host.chooseFolder() },
            modifier = Modifier.fillMaxWidth().heightIn(min = 56.dp),
        ) {
            TextMMD(
                selectedFolderLabel(),
                fontWeight = FontWeight.Bold,
                modifier = Modifier.padding(vertical = 8.dp),
            )
        }
    }

    private fun selectedFolderLabel(): String {
        val stored = prefs.getString(KEY_TREE, null) ?: return "Pictures and Movies / DomonationCamera"
        return try {
            val uri = Uri.parse(stored)
            val documentId = DocumentsContract.getTreeDocumentId(uri)
            when {
                documentId.startsWith("primary:") -> {
                    val relative = documentId.removePrefix("primary:").replace(':', '/')
                    if (relative.isBlank()) "Internal storage" else "Internal storage/$relative"
                }
                documentId.contains(':') -> documentId.replace(':', '/')
                documentId.isNotBlank() -> documentId
                else -> uri.lastPathSegment ?: stored
            }
        } catch (_: Exception) {
            Uri.parse(stored).lastPathSegment ?: stored
        }
    }

    @Composable
    private fun PhotoSettings() {
        Section("Photo")
        ToggleRow("Embed EXIF thumbnail", KEY_THUMB, true)
        Note("The embedded thumbnail improves browsing on devices such as the Canon SELPHY.")
    }

    @Composable
    private fun VideoSettings() {
        Section("Video")
        ToggleRow("Record audio", KEY_VIDEO_AUDIO, true)
        Note("Turn this off for silent video without microphone permission.")
    }

    @Composable
    private fun TimelapseSettings() {
        Section("Timelapse")
        Note("Timelapse frames stay in temporary app storage and are deleted after the video is created. Only the video is saved.")
        HorizontalChoiceRow(
            "Capture interval (seconds)",
            KEY_LAPSE,
            listOf("1" to 1, "2" to 2, "5" to 5, "10" to 10),
            default = 5,
        )
        HorizontalChoiceRow(
            "Playback frame rate (FPS)",
            KEY_LAPSE_FPS,
            listOf("12" to 12, "15" to 15, "24" to 24, "30" to 30),
            default = 24,
        )
    }

    @Composable
    private fun HorizontalChoiceRow(
        label: String,
        key: String,
        choices: List<Pair<String, Int>>,
        default: Int,
    ) {
        val allowed = choices.map { it.second }
        val stored = prefs.getInt(key, default)
        val initial = if (stored in allowed) stored else default
        if (stored != initial) prefs.edit().putInt(key, initial).apply()
        var selected by remember { mutableIntStateOf(initial) }
        TextMMD(label, fontWeight = FontWeight.Bold,
            modifier = Modifier.height(36.dp).wrapContentHeight())
        Row(Modifier.fillMaxWidth().height(52.dp)) {
            choices.forEach { (text, value) ->
                Row(
                    modifier = Modifier.weight(1f).fillMaxHeight().clickable {
                        selected = value
                        prefs.edit().putInt(key, value).apply()
                    },
                    verticalAlignment = Alignment.CenterVertically,
                    horizontalArrangement = Arrangement.Center,
                ) {
                    RadioButtonMMD(selected = selected == value, onClick = {
                        selected = value
                        prefs.edit().putInt(key, value).apply()
                    })
                    TextMMD(text, modifier = Modifier.padding(start = 2.dp))
                }
            }
        }
        HorizontalDividerMMD(color = Color.Black, thickness = 1.dp)
    }

    @Composable
    private fun Section(text: String) {
        Spacer(Modifier.height(8.dp))
        TextMMD(text, fontSize = 19.sp, fontWeight = FontWeight.Bold, modifier = Modifier.height(40.dp).wrapContentHeight())
        HorizontalDividerMMD(color = Color.Black, thickness = 1.dp)
    }

    @Composable
    private fun Note(text: String) {
        TextMMD(text, modifier = Modifier.fillMaxWidth().heightIn(min = 52.dp).wrapContentHeight())
    }

    @Composable
    private fun ToggleRow(label: String, key: String, default: Boolean) {
        var checked by remember { mutableStateOf(prefs.getBoolean(key, default)) }
        Row(
            Modifier.fillMaxWidth().height(56.dp).clickable {
                checked = !checked; prefs.edit().putBoolean(key, checked).apply()
            },
            verticalAlignment = Alignment.CenterVertically,
        ) {
            TextMMD(label, modifier = Modifier.weight(1f))
            SwitchMMD(checked = checked, onCheckedChange = {
                checked = it; prefs.edit().putBoolean(key, it).apply()
            })
        }
        HorizontalDividerMMD(color = Color.Black, thickness = 1.dp)
    }

    @Composable
    private fun ChoiceRow(
        label: String,
        key: String,
        choices: List<Pair<String, Int>>,
        default: Int = choices.first().second,
        onChange: () -> Unit = {},
    ) {
        val allowed = choices.map { it.second }
        val stored = prefs.getInt(key, default)
        val initial = if (stored in allowed) stored else default
        if (stored != initial) prefs.edit().putInt(key, initial).apply()
        var selected by remember { mutableIntStateOf(initial) }
        TextMMD(label, fontWeight = FontWeight.Bold, modifier = Modifier.height(40.dp).wrapContentHeight())
        choices.forEach { (text, value) ->
            Row(
                Modifier.fillMaxWidth().height(52.dp).clickable {
                    selected = value; prefs.edit().putInt(key, value).apply(); onChange()
                },
                verticalAlignment = Alignment.CenterVertically,
            ) {
                RadioButtonMMD(selected = selected == value, onClick = {
                    selected = value; prefs.edit().putInt(key, value).apply(); onChange()
                })
                TextMMD(text, modifier = Modifier.padding(start = 12.dp))
            }
        }
    }

    @Composable
    private fun NumberChoice(key: String, customKey: String, presets: List<Int>, customDefault: Int, min: Int, max: Int) {
        var selected by remember { mutableIntStateOf(prefs.getInt(key, presets.first())) }
        var custom by remember { mutableStateOf(prefs.getInt(customKey, customDefault).toString()) }
        val isCustom = selected !in presets
        Row(Modifier.fillMaxWidth(), horizontalArrangement = Arrangement.SpaceBetween) {
            presets.forEach { value ->
                Row(
                    Modifier.height(52.dp).clickable { selected = value; prefs.edit().putInt(key, value).apply() },
                    verticalAlignment = Alignment.CenterVertically,
                ) {
                    RadioButtonMMD(selected = selected == value, onClick = {
                        selected = value; prefs.edit().putInt(key, value).apply()
                    })
                    TextMMD(value.toString())
                }
            }
            Row(
                Modifier.height(52.dp).clickable {
                    val value = custom.toIntOrNull()?.coerceIn(min, max) ?: customDefault
                    selected = value; prefs.edit().putInt(key, value).putInt(customKey, value).apply()
                },
                verticalAlignment = Alignment.CenterVertically,
            ) {
                RadioButtonMMD(selected = isCustom, onClick = {
                    val value = custom.toIntOrNull()?.coerceIn(min, max) ?: customDefault
                    selected = value; prefs.edit().putInt(key, value).putInt(customKey, value).apply()
                })
                TextMMD("Custom")
            }
        }
        if (isCustom) {
            TextFieldMMD(
                value = custom,
                onValueChange = { input ->
                    if (input.all(Char::isDigit)) {
                        custom = input
                        input.toIntOrNull()?.takeIf { it in min..max }?.let {
                            selected = it; prefs.edit().putInt(key, it).putInt(customKey, it).apply()
                        }
                    }
                },
                modifier = Modifier.fillMaxWidth(),
                label = { TextMMD("$min–$max") },
                keyboardOptions = KeyboardOptions(keyboardType = KeyboardType.Number),
                singleLine = true,
            )
        }
    }

    companion object {
        private const val KEY_TREE = "save_tree"
        private const val KEY_REVIEW = "review_before_save"
        private const val KEY_THUMB = "exif_thumbnail"
        private const val KEY_VIDEO_AUDIO = "video_audio"
        private const val KEY_VOLUME_SHUTTER = "volume_shutter"
        private const val KEY_FLIP_FRONT_CAMERA = "flip_front_camera"
        private const val KEY_LAPSE = "timelapse_seconds"
        private const val KEY_LAPSE_FPS = "timelapse_fps"
    }
}
